tchannel/peer_heap.py

```diff
@@ -0,0 +1,157 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+
+import random
+
+from .container import heap
+from .container.heap import HeapOperation
+from .container.heap import NoMatchError
+from .container.heap import smallest
+
+
+class PeerHeap(HeapOperation):
+    """PeerHeap maintains a min-heap of peers based on their scores.
+
+    Peer in the heap will be arranged based on the peer's score and peer's
+    order. Order is equal to peer heap's current order number plus random value
+    within the current peer size. It solves two problems when peers are in
+    same score:
+
+    Dead peers: If the order is completely random, then an unlucky peer
+    with a very bad assigned order may never get selected.
+
+    Determinism: If the insertion order is used as-is, then all TChannel
+    instances would follow the same selection pattern, causing load
+    imbalance. For example if they get the same static list of peers, they
+    will all pick the first one, then the second one, and so on, cycling
+    between which host gets overloaded with requests.
+
+    All in all, it will keep certain level randomization but at the same
+    time make the peer score not deterministic among different tchannel
+    instances.
+
+    """
+
+    __slots__ = ('peers',)
+
+    def __init__(self):
+        self.peers = []
+        self.order = 0
+
+    def size(self):
+        return len(self.peers)
+
+    def lt(self, i, j):
+        """Compare the priority of two peers.
+
+        Primary comparator will be the score of each peer. If the ``score`` is
+        same then compare the ``order``. The ``order`` attribute of the peer
+        tracks the heap push order of the peer. This help solve the imbalance
+        problem caused by randomization when deal with same score situation.
+
+        :param i: ith peer
+        :param j: jth peer
+        :return: True or False
+        """
+        if self.peers[i].score == self.peers[j].score:
```

Comment:
I mentioned this in the original PR too: Calling this score is really confusing because that implies that high score means the peer is more desirable, while the behavior we are actually using is that low score means that the peer is more desirable. I've already been tripped by that in the previous code reviews. I highly recommend renaming `score` to something that implies that lower is better. You mentioned that `pending`/`ongoing` (as in the number of ongoing requests) didn't work here because we'll eventually want more complex criteria for measuring the score. How about `rank`? Or we could still call it `ongoing` or `pending` and rename it if and when we add more complex ranking criteria?